### PR TITLE
add third party thread pool

### PR DESCRIPTION
**Pull request by astrophysik**
General
----------
Main goal of this pr is to add simple third-party thread pool to test some hypotheses. The implementation from [here](https://github.com/bshoshany/thread-pool) was chosen as the baseline. In the future, this [library](https://yaclib.github.io/YACLib/ ) may be chosen.

Details
---------
An option `thread-pool-ratio` has been added to define the number of threads. That define threads count ratio of the overall cpu numbers

Also this pr add simple parallel function `do_parallel_job(...)` to check if the system is working well

### CODE REVIEW

File: runtime/thread-pool.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2023 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "runtime/thread-pool.h"
+
+void ThreadPool::init() noexcept {
+  int thread_pool_size = static_cast<int>(std::thread::hardware_concurrency() * thread_pool_ratio);
```
Comment (DrDet) [inline runtime/thread-pool.cpp:8]: let's round up using ceil to ensure it always `> 0` if you specify non-zero ratio

File: runtime/interface.cpp [hunk 2]
```diff
@@ -2370,6 +2395,8 @@ static void free_interface_lib() {
 static void free_runtime_libs() {
   php_assert (dl::in_critical_section == 0);
 
+  vk::singleton<ThreadPool>::get().stop();
```
Comment (DrDet) [inline runtime/interface.cpp:2398]: I think it's redundant

File: runtime/interface.cpp [hunk 3]
```diff
@@ -2345,6 +2368,8 @@ static void init_runtime_libs() {
   init_string_buffer_lib(static_cast<int>(static_buffer_length_limit));
 
   init_interface_lib();
+
+  vk::singleton<ThreadPool>::get().init();
```
Comment (DrDet) [inline runtime/interface.cpp:2372]: let's move it to `worker_global_init`
https://github.com/VKCOM/kphp/blob/a9aafd0bf159fa43675253fe336ddebad2b4943b/runtime/interface.h#L209

File: server/server-stats.cpp [hunk 4]
```diff
@@ -475,6 +476,7 @@ struct WorkerProcessStats : private vk::not_copyable {
     misc_stats.inc_stat(MiscStat::Key::worker_activity_counter, worker_index);
     misc_stats.set_stat(MiscStat::Key::json_logs_count, worker_index, vk::singleton<JsonLogger>::get().get_json_logs_count());
     misc_stats.set_stat(MiscStat::Key::json_traces_count, worker_index, vk::singleton<JsonLogger>::get().get_json_traces_count());
+    misc_stats.set_stat(MiscStat::Key::threads_count, worker_index, get_self_threads_count());
   }
```
Comment (DrDet) [inline server/server-stats.cpp:481]: you forgot to reset new stat here:
https://github.com/VKCOM/kphp/blob/a9aafd0bf159fa43675253fe336ddebad2b4943b/server/server-stats.cpp#L493

**Review by DrDet — CHANGES_REQUESTED** [review]

File: runtime/thread-pool.cpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * linux does not determinize the thread to which the signal will arrive,
+     * for this purpose runtime signals is blocked at thread creation time. Since they inherit the signaling mask
+     * */
+    sigset_t sigset = dl_get_empty_sigset();
+    sigaddset(&sigset, SIGALRM);
+    sigaddset(&sigset, SIGUSR1);
+    sigaddset(&sigset, SIGUSR2);
+    sigaddset(&sigset, SIGRTMIN + 1);
+    sigaddset(&sigset, SIGRTMIN + 2);
```
[the remark this answers is not in the document]
Comment (astrophysik) [reply runtime/thread-pool.cpp:19]: fix

File: server/server-stats.cpp (hunk 4 above)
Comment (astrophysik) [reply server/server-stats.cpp:481]: done

File: runtime/interface.cpp (hunk 3 above)
Comment (astrophysik) [reply runtime/interface.cpp:2372]: done

**Review by DrDet — APPROVED** [review]

**Review by DrDet — APPROVED** [review]

**Pull request merged**